npm_and_yarn/lib/dependabot/npm_and_yarn/package_manager.rb

```diff
@@ -10,18 +11,52 @@ def initialize(package_json, lockfiles:)
         @package_json = package_json
         @lockfiles = lockfiles
         @package_manager = package_json.fetch("packageManager", nil)
+        @engines = package_json.fetch("engines", nil)
       end
 
+      # rubocop:disable Metrics/CyclomaticComplexity
+      # rubocop:disable Metrics/PerceivedComplexity
       def setup(name)
-        return unless @package_manager.nil? || @package_manager.start_with?("#{name}@")
+        # we prioritize version mentioned in "packageManager" instead of "engines"
+        # i.e. if { engines : "pnpm" : "6" } and { packageManager: "pnpm@6.0.2" },
+        # we go for the sepcificity mentioned in packageManager (6.0.2)
+        # Dependabot::Experiments.register(:enable_pnpm_yarn_dynamic_engine, true)
+        if Dependabot::Experiments.enabled?("enable_pnpm_yarn_dynamic_engine")
+
+          unless @package_manager.nil? || @package_manager.start_with?("#{name}@") ||
+                 (@package_manager && @package_manager == name.to_s)
```

Comment:
nit: this is hard to read, but I think it can be simplified:

```suggestion
          unless @package_manager&.start_with?("#{name}@") || @package_manager == name.to_s)
```

Reply:
@JamieMagee , i have kept _nil?_ check for regression (makes easier to revert back in case). I have applied safe nav operators.